bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/websocket/TibberWebsocket.java

```diff
@@ -0,0 +1,243 @@
+                logger.warn("Exception getting ws url {}", e.getMessage());
+            }
+        }
+        return null;
+    }
+
+    public void stop() {
+        wsClient.ifPresentOrElse(client -> {
+            try {
+                session.ifPresent(session -> {
+                    sendMessage(DISCONNECT_MESSAGE);
+                    session.close();
+                });
+                client.stop();
+                client.destroy();
+            } catch (Exception e) {
+                logger.warn("Exception stopping ws url {}", e.getMessage());
+            } finally {
+                wsClient = Optional.empty();
+                session = Optional.empty();
+            }
+        }, () -> {
+            wsClient = Optional.empty();
+            session = Optional.empty();
+        });
+    }
+
+    @OnWebSocketConnect
+    public void onConnect(Session wssession) {
+        session = Optional.of(wssession);
+        String connection = String.format(CONNECT_MESSAGE, config.token);
+        sendMessage(connection);
+    }
+
+    @OnWebSocketClose
+    public void onClose(int statusCode, String reason) {
+        session = Optional.empty();
+        logger.debug("WebSocket closed - Status {} Reason {}", statusCode, reason);
+    }
+
+    @OnWebSocketError
+    public void onWebSocketError(Throwable e) {
+        session = Optional.empty();
+        logger.warn("Websocket error {}", e.getMessage());
+    }
+
+    @OnWebSocketMessage
+    public void onMessage(String message) {
+        if (message.contains("connection_ack")) {
+            logger.debug("WebSocket connected to Server");
+            // String subScriptionMessage = String.format(handler.getTemplate(WEBSOCKET_SUBSCRIPTION_RESOURCE_PATH),
+            // config.homeid);
+            String subScriptionMessage = String.format(SUBSCRIPTION_MESSAGE, config.homeid);
+            sendMessage(subScriptionMessage);
+        } else {
+            handler.newMessage(message);
+        }
+    }
+
+    @OnWebSocketFrame
+    public void onFrame(Frame frame) {
+        if (Frame.Type.PONG.equals(frame.getType())) {
+            ByteBuffer buffer = frame.getPayload();
+            byte[] bytes = new byte[frame.getPayloadLength()];
+            for (int i = 0; i < frame.getPayloadLength(); i++) {
+                bytes[i] = buffer.get(i);
+            }
+            String paylodString = new String(bytes);
+            Instant sent = pingPongMap.remove(paylodString);
+            if (sent == null) {
+                logger.debug("Websocket receiced pong without ping {}", paylodString);
+            }
+        } else if (Frame.Type.PING.equals(frame.getType())) {
+            session.ifPresentOrElse((session) -> {
+                ByteBuffer buffer = frame.getPayload();
+                try {
+                    session.getRemote().sendPong(buffer);
+                } catch (IOException e) {
+                    logger.warn("Websocket onPing answer exception {}", e.getMessage());
```

Comment:
What is the consequence of this happening? Will the user need to take some action? Consider log level.